File: android/app/src/main/java/com/lttprandomizer/RomUtils.kt
package com.lttprandomizer

/**
 * Shared ROM utilities used by both BpsPatcher and CosmeticPatcher.
 */
object RomUtils {
    /**
     * Recalculates and writes the SNES HiROM checksum at 0x7FDC–0x7FDF.
     * Modifies [rom] in place.
     */
    fun writeChecksum(rom: ByteArray) {
        rom[0x7FDC] = 0; rom[0x7FDD] = 0; rom[0x7FDE] = 0; rom[0x7FDF] = 0

        var sum = 0L
        for (b in rom) sum += (b.toInt() and 0xFF)
        val checksum   = (sum and 0xFFFFL).toInt()
        val complement = checksum xor 0xFFFF

        rom[0x7FDC] = (complement and 0xFF).toByte()
        rom[0x7FDD] = (complement shr 8).toByte()
        rom[0x7FDE] = (checksum and 0xFF).toByte()
        rom[0x7FDF] = (checksum shr 8).toByte()
    }
}

File: windows/Services/RomUtils.cs
namespace LTTPRandomizerGenerator.Services
{
    /// <summary>
    /// Shared ROM utilities used by both BpsPatcher and CosmeticPatcher.
    /// </summary>
    public static class RomUtils
    {
        /// <summary>
        /// Recalculates and writes the SNES HiROM checksum at 0x7FDC–0x7FDF.
        /// </summary>
        public static void WriteChecksum(byte[] rom)
        {
            rom[0x7FDC] = rom[0x7FDD] = rom[0x7FDE] = rom[0x7FDF] = 0;

            uint sum = 0;
            foreach (byte b in rom) sum += b;
            ushort checksum   = (ushort)(sum & 0xFFFF);
            ushort complement = (ushort)(checksum ^ 0xFFFF);

            rom[0x7FDC] = (byte)(complement & 0xFF);
            rom[0x7FDD] = (byte)(complement >> 8);
            rom[0x7FDE] = (byte)(checksum & 0xFF);
            rom[0x7FDF] = (byte)(checksum >> 8);
        }
    }
}
